diff.py

```diff
@@ -1880,21 +1950,136 @@ def progress(self, msg: str) -> None:
         sys.stdout.write("\x1b7\x1b[1;1f{}\x1b8".format(msg + " "))
         sys.stdout.flush()
 
-    def update(self, text: str, error: bool) -> None:
-        if not error and not self.emsg and text == self.mydump:
+    def update(self, mydump: str) -> None:
+        if not self.emsg and mydump == self.mydump:
             self.progress("Unchanged. ")
             return
-        self.pending_update = (text, error)
-        if not self.less_proc:
+        self.pending_update = (mydump, False)
+        self.kill_less()
+
+    def update_error(self, error: str) -> None:
+        self.pending_update = (error, True)
+        self.kill_less()
+
+    def terminate(self) -> None:
+        self.pending_update = None
+        self.kill_less()
+
+
+class WebDisplay(Display):
+    watch_queue: "queue.Queue[Optional[float]]"
+    ready_queue: "queue.Queue[Optional[float]]"
+    http_server: http.server.HTTPServer
+    running: bool
+
+    def run_sync(self) -> None:
+        web_display = self
+
+        class MyReqHandler(http.server.BaseHTTPRequestHandler):
+            def do_GET(self):
+                if not web_display.running:
+                    return
+                self.send_response(http.HTTPStatus.OK)
+                self.send_header("Access-Control-Allow-Origin", "*")
+                self.end_headers()
+                self.wfile.write(web_display.run_diff().encode("utf-8"))
+                self.wfile.flush()
+                # fixme
+                threading.Thread(target=web_display.http_server.shutdown).start()
+
+        self.http_server = http.server.HTTPServer(
+            ("localhost", self.config.http_server_port), MyReqHandler
+        )
+        self.running = True
+        # fixme
+        print(f"Http server is running on port {self.config.http_server_port}")
+        print("Open with ?once (eg file:///.../client.html?once)")
```

Comment:
file:/// isn't great; it's generally low-privileged and behaves differently between browsers wrt what files it can access. would be better if we could serve it from the same port

Reply:
You mean serve client.html through diff.py ? good idea